dbms/src/Functions/FunctionsStringSearch.h

```diff
@@ -16,13 +16,118 @@
+#include <Common/Exception.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/UTF8Helpers.h>
+#include <Common/typeid_cast.h>
 #include <DataTypes/DataTypeString.h>
 #include <DataTypes/DataTypesNumber.h>
 #include <Functions/FunctionHelpers.h>
 #include <Functions/IFunction.h>
+#include <Functions/StringUtil.h>
+
+#include <cstring>
+#include <string_view>
 
 namespace DB
 {
+using Chars_t = ColumnString::Chars_t;
+using Offsets = ColumnString::Offsets;
+
+inline void copyColumnStringDataImpl(ColumnString * dst_col, const ColumnString * src_col)
+{
+    auto & dst_data = dst_col->getChars();
+    auto & dst_offsets = dst_col->getOffsets();
+
+    const auto & src_data = src_col->getChars();
+    const auto & src_offsets = src_col->getOffsets();
+
+    dst_data.resize(src_data.size());
+    dst_offsets.resize(src_offsets.size());
+
+    memcpy(&dst_data[0], &src_data[0], src_data.size());
+    memcpy(&dst_offsets[0], &src_offsets[0], src_offsets.size() * sizeof(ColumnString::Offset));
+}
+
+inline void copyColumnStringData(MutableColumnPtr & dst_col, const ColumnPtr & src_col, const ColumnConst * src_col_const)
+{
+    if (src_col_const != nullptr)
+    {
+        auto dst_str_col = ColumnString::create();
+        copyColumnStringDataImpl(&*dst_str_col, typeid_cast<const ColumnString *>(&src_col_const->getDataColumn()));
+        dst_col = ColumnConst::create(ColumnPtr(std::move(dst_str_col)), src_col_const->size());
+    }
+    else
+    {
+        dst_col = ColumnString::create();
+        copyColumnStringDataImpl(typeid_cast<ColumnString *>(&*dst_col), typeid_cast<const ColumnString *>(&*src_col));
+    }
+}
+
+inline void lowerOneString(Chars_t & str_container, size_t pos, size_t size)
+{
+    size_t i = 0;
+    while (i < size)
+    {
+        if (isASCII(str_container[pos + i]))
+        {
+            if (isUpperAlphaASCII(str_container[pos + i]))
+                str_container[pos + i] = toLowerIfAlphaASCII(str_container[pos + i]);
+            ++i;
+        }
+        else
+        {
+            size_t utf8_len = UTF8::seqLength(str_container[pos + i]);
+            i += utf8_len;
+        }
+    }
+}
+
+inline void lowerColumnConst(ColumnConst * lowered_col_const)
+{
+    auto * col_data = typeid_cast<ColumnString *>(&lowered_col_const->getDataColumn());
+    RUNTIME_ASSERT(col_data != nullptr, "Invalid column type, should be ColumnString");
+
+    auto & str_container = col_data->getChars();
+    auto str_size = col_data->getOffsets()[0];
+
+    lowerOneString(str_container, 0, str_size);
+}
+
+inline void lowerColumnString(MutableColumnPtr & col)
+{
+    auto * col_vector = typeid_cast<ColumnString *>(&*col);
+    RUNTIME_ASSERT(col_vector != nullptr, "Invalid column type, should be ColumnString");
+
+    auto & str_container = col_vector->getChars();
+    const auto & str_offsets = col_vector->getOffsets();
+
+    size_t str_num = col_vector->size();
+    for (size_t i = 0; i < str_num; ++i)
```

Comment:
> I think it we can just call lowerOneString(str_container, 0, StringUtil::sizeAt(str_offsets, str_offsets.size()))?

This will check many non-ascii characters and it's unnecessary to do this.

Reply:
Yes, when it's ci collator, the non-ascii will not be lowered.

Reply:
> > I think it we can just call lowerOneString(str_container, 0, StringUtil::sizeAt(str_offsets, str_offsets.size()))?
> 
> This will check many non-ascii characters and it's unnecessary to do this.

I don't get your point, the number of characters veen checked should be the same

Reply:
> > > I think it we can just call lowerOneString(str_container, 0, StringUtil::sizeAt(str_offsets, str_offsets.size()))?
> > 
> > 
> > This will check many non-ascii characters and it's unnecessary to do this.
> 
> I don't get your point, the number of characters veen checked should be the same

sry, I misunderstand